python/ray/serve/_private/application_state.py

```diff
@@ -334,50 +403,89 @@ def _determine_app_status(self) -> Tuple[ApplicationStatus, str]:
         else:
             return ApplicationStatus.DEPLOYING, ""
 
-    def _reconcile_deploy_obj_ref(self) -> Tuple[BuildAppStatus, str]:
-        """Reconcile the in-progress deploy task.
+    def _start_or_reconcile_build_app_task(
+        self, code_version: str
+    ) -> Tuple[List, BuildAppStatus, str]:
+        """Start or reconcile the in-progress build task.
 
-        Resets the self._deploy_obj_ref if it finished, regardless of
-        whether it finished successfully.
+        If there is no in-progress build task with the correct code
+        version, start a new one. Otherwise, reconcile the in-progress
+        build task (resets self._build_app_task_info when task finishes,
+        regardless of whether it finished successfully)
```

Comment:
Related to my comment below: what's the purpose of resetting `self._build_app_task_info`? Couldn't we instead mark it done and then short-circuit this method if the version matches the target version?

Might be missing this being handled at a higher layer somewhere.